### PR TITLE
[algo, ci] feat: add SAPO loss, example and tests

### PR DESCRIPTION
**Pull request by CedricHwong**
### What does this PR do?

Issue: #4621 

  This PR adds the SAPO policy loss with a recipe/config to enable it, and adds a CPU unit test to validate SAPO loss gating, masking, and KL metrics. It ensures SAPO loss behavior is covered by tests and ready for use in training.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

- pytest tests/trainer/ppo/test_core_algos_on_cpu.py -k sapo

### API and Usage Example

  > No public API changes. Example configuration tweak:
  
  ```bash
  # Enable SAPO loss in a trainer config object
  actor_rollout_ref.actor.policy_loss.loss_mode = "sapo"
  actor_rollout_ref.actor.policy_loss.tau_pos = 1.0
  actor_rollout_ref.actor.policy_loss.tau_neg = 1.05
  ```
### Design & Code Changes

  > - Add SAPO policy loss with soft gating in verl/trainer/ppo/core_algos.py.
  > - Add CPU unit test for SAPO loss in tests/trainer/ppo/test_core_algos_on_cpu.py.
  > - Remove recipe/sapo (entrypoint/config/run script).
  > - Add examples/sapo_trainer/run_qwen2-7b.sh that invokes verl.trainer.main_ppo with SAPO
    loss mode and tau_pos/tau_neg.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [x] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: recipe/sapo/config/sapo_trainer.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+hydra:
```
Comment (vermouth1992) [inline recipe/sapo/config/sapo_trainer.yaml]: Can we run sapo from main directly instead of creating a new recipe?

Comment (CedricHwong) [reply recipe/sapo/config/sapo_trainer.yaml]: Sure，will do sir🫡

**Comment by vermouth1992** [comment]
It seems that there are two SAPO recipes. Shall we merge them into one and acknowledge both contribution? https://github.com/volcengine/verl/pull/4345

**Comment by CedricHwong** [comment]
@vermouth1992  @BounharAbdelaziz

This update aligns SAPO with the paper’s exact formulation: the gate uses f_i,t(x)=σ(τ_i,t(x−1)) * (4/τ_i,t) with τ chosen by advantage sign, and the loss is aggregated at the sequence level (seq-mean-token-mean) per Eq. (5). The rest of the loss flow is unchanged (ratio and KL tracking), but the gradient scaling and sequence coherence now match the paper.

Comparison With PR #4345  

  - Gate formula: #4345 matched the paper; my original ipml. used a plain sigmoid. This update
    makes #4624 match the paper exactly.
  - Aggregation: Bounhar enforced seq-mean-token-mean; my original ipml. defaulted to token-mean. This update enforces the paper’s token‑adaptive gating + sequence‑level objective aggregation.
  - Config structure: my implementation keeps tau_pos/tau_neg under policy_loss (cleaner and consistent);


Recommendation to Maintainers

Merge the updated #4624 as the canonical SAPO implementation: it now matches the paper’s math, keeps configuration clean, and preserves test coverage. If useful, cherry-pick any large-scale training scripts from #4345 separately.

Credit

The paper-aligned SAPO gate implementation is credited to @BounharAbdelaziz (added as co-author via GitHub noreply).

**Comment by vermouth1992** [comment]
Why did this PR close?

**Comment by CedricHwong** [comment]
@vermouth1992 Cause I believe he @BounharAbdelaziz has already done a very good SAPO implementation. Perhaps I should check more carefully to see if there are any PRs with similar purposes. GitHub PR searches can be unpredictable sometimes, haha.
